pt_miniscreen/pages/root/projects.py

```diff
@@ -64,86 +68,190 @@ def from_file(cls, file):
 
-    return env
+            if exit_condition == ProjectExitCondition.POWER_BUTTON_PRESS:
+                event_callback = {Message.PUB_V3_BUTTON_POWER_PRESSED: self.stop}
+            elif exit_condition == ProjectExitCondition.HOLD_X:
+                CANCEL_BUTTON_PRESS_TIME = 3
 
+                timer = Timer(CANCEL_BUTTON_PRESS_TIME, self.stop)
 
-def run_project(cmd: str, cwd: str = None):
-    logger.info(f"run_project(command_str='{cmd}', cwd='{cwd}')")
-    user = get_user_using_first_display()
-    env = get_project_environment()
+                def on_cancel_button_pressed():
+                    nonlocal timer
+                    timer.cancel()
+                    timer = Timer(CANCEL_BUTTON_PRESS_TIME, self.stop)
+                    timer.start()
 
-    return Popen(
-        split(cmd),
-        env=env,
-        cwd=cwd,
-        preexec_fn=lambda: switch_user(user),
-    )
+                def on_cancel_button_release():
+                    nonlocal timer
+                    timer.cancel()
+
+                event_callback = {
+                    Message.PUB_V3_BUTTON_CANCEL_PRESSED: on_cancel_button_pressed,
+                    Message.PUB_V3_BUTTON_CANCEL_RELEASED: on_cancel_button_release,
+                }
+
+            self.subscribe_client = PTDMSubscribeClient()
+            self.subscribe_client.initialise(event_callback)
+            self.subscribe_client.start_listening()
+            logger.info(f"Using exit condition {exit_condition.name}")
+        except Exception:
+            logger.info("Not using an exit condition")
+
+
+class ProjectState(Enum):
+    IDLE = auto()
+    STARTING = auto()
+    RUNNING = auto()
+    STOPPING = auto()
+    ERROR = auto()
 
 
 class ProjectPage(Component):
     def __init__(self, project_config: ProjectConfig, **kwargs):
         self.project_config = project_config
+        self.state = ProjectState.IDLE
         super().__init__(**kwargs)
 
-        self.process = None
         self.text = self.create_child(
-            MarqueeText,
-            text=f"Running '{self.project_config.title}'",
-            font_size=14,
+            Text,
+            text=self.displayed_text,
+            get_text=lambda: self.displayed_text,
+            font_size=10,
             align="center",
             vertical_align="center",
         )
 
-    def run(self, on_stop: Callable = None):
-        self.start()
-
-        def wait_and_run_on_stop():
-            self.wait()
-            on_stop()
+    @property
+    def displayed_text(self) -> str:
+        text = ""
+        if self.state == ProjectState.ERROR:
+            text = f"There was an error while running '{self.project_config.title}'..."
+        elif self.state == ProjectState.STOPPING:
+            text = f"Stopping '{self.project_config.title}'..."
+        elif self.state == ProjectState.STARTING:
+            text = f"Starting '{self.project_config.title}'..."
```

Comment:
There should also be text for when the project is running since there is no guarantee that user projects take over the miniscreen.

```suggestion
        elif self.state == ProjectState.STARTING:
            text = f"Starting '{self.project_config.title}'..."
        elif self.state == ProjectState.RUNNING:
            text = f"Running '{self.project_config.title}'..."
```